lib/charms/prometheus_k8s/v1/prometheus.py

```diff
@@ -208,66 +189,37 @@ def __init__(self, charm, name, service, version=None):
         """
         super().__init__(charm, name, service, version)
         self._charm = charm
-        self._stored.set_default(jobs={})
+        self._relation_name = name
         events = self._charm.on[name]
         self.framework.observe(
             events.relation_changed, self._on_scrape_target_relation_changed
         )
         self.framework.observe(
-            events.relation_broken, self._on_scrape_target_relation_broken
+            events.relation_departed, self._on_scrape_target_relation_departed
         )
 
     def _on_scrape_target_relation_changed(self, event):
         """Handle changes in related consumers.
 
         Anytime there are changes in relations between Prometheus
-        provider and consumer charms the scrape job config is updated
-        and the Prometheus charm is informed, through a
-        `TargetsChanged` event. The Prometheus charm can then
-        choose to update its scrape configuration.
+        provider and consumer charms the Prometheus charm is informed,
+        through a `TargetsChanged` event. The Prometheus charm can
+        then choose to update its scrape configuration.
         """
-        if not self._charm.unit.is_leader():
-            return
-
         rel_id = event.relation.id
-        data = event.relation.data[event.app]
 
-        targets = json.loads(data.get("targets", "[]"))
-        if not targets:
-            return
+        self.on.targets_changed.emit(relation_id=rel_id)
 
-        job_name = data.get("job_name", "")
-        unique_name = "juju_{}_{}_{}".format(
-            self._charm.model.name, self._charm.app.name, rel_id
-        )
-        if job_name:
-            job_name += "_{}".format(unique_name)
-        else:
-            job_name = unique_name
-
-        job_config = {"job_name": job_name, "static_configs": [{"targets": targets}]}
-
-        self._stored.jobs[rel_id] = json.dumps(job_config)
-        logger.debug("New job config on relation change : %s", job_config)
-        self.on.targets_changed.emit()
-
-    def _on_scrape_target_relation_broken(self, event):
+    def _on_scrape_target_relation_departed(self, event):
         """Update job config when consumers depart.
 
         When a Prometheus consumer departs the scrape configuration
-        for that consumer is remove from the list of scrape jobs and
+        for that consumer is removed from the list of scrape jobs and
         the Prometheus is informed through a `TargetsChanged`
         event.
         """
-        if not self._charm.unit.is_leader():
-            return
-
         rel_id = event.relation.id
-        try:
-            del self._stored.jobs[rel_id]
-            self.on.targets_changed.emit()
-        except KeyError:
-            pass
+        self.on.targets_changed.emit(relation_id=rel_id)
```

Comment:
This is risky, for the same reason as the Karma charm from @sed-i. Even after `RelationDeparted` or `RelationBroken`, the data bag still exists. I would suspect that, even after `RelationDeparted` triggers this event, the list of jobs populated when the charm calls `provider.jobs()` will include this despite it being a broken relation (which has not finished departing yet)

Reply:
Thanks. I am not seeing this behaviour. When a unit is removed Prometheus dashboard does reflect this change. Will test again.

Reply:
Tested again. Units do get removed from Prometheus scrape configuration when they depart. This was tested using the [prometheus-tester](git@github.com:balbirthomas/prometheus-tester.git) charm. Do let me know if there any other way you would like me to test or if there is anything else I can try.

If a unit departs and its *unit data bag* still persists or if and application is removed or its relation removed and the *application data bag* still persists, I would presume this would be considered a bug in Juju or the Operator Framework. Even though I am not seeing this, do let me know if I am mistaken in this presumption.

Leaving as is for now.

Reply:
So the basic workflow @sed-i has is the same. The events are not actually asynchronous (except where they're coming from OF). The data bags still exist, at least for applications, until `RelationDeparted` or `RelationBroken` finishes. At least walking the Juju code (which I'm not that familiar with yet), and from @sed-i testing.

With a call stack more or less like:

```
ENTER relation_departed
  ENTER _on_scrape_relation_departed
  EMIT targets_changed
    ENTER charm.on_targets_changed
      ENTER charm._configure
        ENTER provder.jobs()
        # databag still exists, relation_departed has not finished
      EXIT
    EXIT
  EXIT
EXIT
```

Units do get removed from the scrape configuration. From your patch, I wonder if _applications_ do. If an arbitrary, say, Cassandra/MySQL enters a scrape relation, then the application leaves entirely, we would see `provder.jobs()` "successfully" iterate over the relations (including that one), "successfully" get the scrape targets from the databag, and set it in the config.

I did not test @sed-i 's charm myself, but now that it's a similar conversation two days in a row, I'll build from this branch and test for sanity.

Reply:
Applications also do get removed from the data bag. AFAIU the workflow is 

ENTER relation_departed
  ENTER _on_scrape_relation_departed
  EMIT targets_changed
  EXIT
EXIT
# Juju fires targets_changed

Reply:
Tested again. Both `remove-relation` and `remove-application` results in scrape target being removed from Prometheus dashboard. Let me know if there is any other test you would like me to try.

Reply:
I can confirm metadata was remaining in `prometheus.yml`. I have pushed a [commit](https://github.com/canonical/prometheus-operator/pull/71/commits/6ed46985c2ab2ac6eb382b08651413ddc527959a) fixing this. I did check this seems to have fixed the issue. Thanks.